File: Arrays & Hashing/productExceptSelf.go
// Given an integer array nums, return an array output where output[i] is the product of all the elements of nums except nums[i].
// Each product is guaranteed to fit in a 32-bit integer.
// https://leetcode.com/problems/products-of-array-except-self/

package main

import (
	"fmt"
)

func productExceptSelfV2(nums []int) []int {
	res := make([]int, len(nums))
	prefix := 1
	for i := 0; i < len(nums); i++ {
		res[i] = prefix
		prefix *= nums[i]
	}
	postfix := 1
	for i := len(nums) - 1; i >= 0; i-- {
		res[i] *= postfix
		postfix *= nums[i]
	}
	return res
}

func productExceptSelf(nums []int) []int {
	output := make([]int, len(nums))
	for i := 0; i < len(nums); i++ {
		product := 1
		for j := 0; j < len(nums); j++ {
			if i == j {
				continue
			}
			product *= nums[j]
		}
		output[i] = product
	}
	return output
}

func main() {
	nums := []int{1, 2, 4, 6}
	fmt.Println(productExceptSelf(nums))
	fmt.Println(productExceptSelfV2(nums))
}

File: Arrays & Hashing/productExceptSelf.py
# Given an integer array nums, return an array output where output[i] is the product of all the elements of nums except nums[i].
# Each product is guaranteed to fit in a 32-bit integer.
# https://leetcode.com/problems/products-of-array-except-self/

class Solution:
    def productExceptSelfV2(self, nums):
        res = [1] * len(nums)
        prefix = 1
        for i in range(len(nums)):
            res[i] = prefix
            prefix *= nums[i]
        postfix = 1
        for i in range(len(nums) - 1, -1, -1):
            res[i] *= postfix
            postfix *= nums[i]
        return res

    def productExceptSelf(self, nums):
        output = []
        for i in range(len(nums)):
            product = 1;
            for j in range(len(nums)):
                if (i == j):
                    continue
                product *= nums[j]

            output.append(product)
        return output


def main():
    nums = [1, 2, 4, 6]
    print(Solution().productExceptSelf(nums))
    print(Solution().productExceptSelfV2(nums))

if __name__ == "__main__":
    main()
